docs/proposals/scheduling/20220629-schedule-device-in-card-level.md

```diff
@@ -0,0 +1,355 @@
+}
+
+type DeviceInfo struct {
+    ID        string                       `json:"id"`       // Device UUID
+    Resources map[string]resource.Quantity `json:"resource"` // map of sub-resources
+    Health    bool                         `json:"health"`
+    Minor     int32                        `json:"minor"` // index starting from 0
+}
+
+type DeviceTopology struct {
+	Links [][]DeviceLink `json:"links,omitempty"`
+}
+type Device struct {
+	Type      string     `json:"gpuLinkType,omitempty"`
+	Bandwidth int64      `json:"bandwidth,omitempty"`
+}
+```
+
+#### RuntimeProxy
+Our target is to work compatible with origin k8s kubelet and k8s device plugins, so:
+
+1.We still allow kubelet and device plugin to allocate concrete device card. which means no matter there's a k8s device 
+plugin or not, our design can work well.
+
+2.In runtimeproxy, we will use pod's devive-allocation-result in annotation to replace the step1's result of container's 
+args and envs.
+
+We should modify protocol between runtimeproxy and koordlet to add container env:
+```go
+type ContainerResourceHookResponse struct {
+    state         protoimpl.MessageState
+    sizeCache     protoimpl.SizeCache
+    unknownFields protoimpl.UnknownFields
+
+    ContainerAnnotations map[string]string        
+    ContainerResources   *LinuxContainerResources 
+    PodCgroupParent      string   
+    
+    //new added
+    ENV map[string]string
+}
+```
+
+Then we will add a new gpu-hook in koordlet's runtimehooks. It's used in PreStartContainer stage. We will generate new gpu 
+env(NVIDIA_VISIBLE_DEVICES) by pod gpu allocation info in annotation.
+
+Finally, we should modify ContainerResourceExecutor's UpdateRequest function in runtimeproxy, and let new gpu-env covering
+old gpu-env.
+
+### Compatibility
+As we know, the gpu scheduling in kube-scheduler side has no any different with other scalar resources. The concrete 
+card-level assigning is done by kubelet and gpu device plugin, which will generate container's gpu env. 
+
+Our design has no conflict with the above process. our device reporter will report koordinator gpu resources for kubelet
+updating node resources. Then we schedule device request in our new plugin with new device resource account. In pre-bind 
+stage, we will update container resources with koordinator gpu resources, this is for kubelet to check resource limitation.
+We will also add device allocation information to pod's annotation. In node side, the k8s device plugin will first patch
+container env, but we will overwrite these envs in runtimeproxy by allocation result in pod's annotation.
+
+If there is no gpu device plugin, the only different is in pre-bind stage, we will erase the origin-gpu-resource-dimension
+in container to pass kubelet resource limitation check. This will be a flag in scheduler side.
+
+When we hot-update, how can we handle the scheduled-pod without device allocation in pod's annotation? We can get 
```

Comment:
For those pods whose GPUs are already allocated by kubelet(before koordinator installation), koordlet also needs to report the allocs in CRD, so that scheduler can use them for making allocations.

Reply:
as we discuss offline, we can get the allocation information for container's env to help recovering device cache usage